cpp/src/arrow/builder.h

```diff
@@ -179,13 +180,43 @@ class ARROW_EXPORT PrimitiveBuilder : public ArrayBuilder {
   value_type* raw_data_;
 };
 
+template <typename T>
+struct BuilderTypeTraits {
+  constexpr static bool has_simple_type = false;
+};
+
+#define SIMPLE_BUILDER_TRAITS(ArrowType, Constructor)                                    \
+  template <>                                                                            \
+  struct BuilderTypeTraits<ArrowType> {                                                  \
+    constexpr static bool has_simple_type = true;                                        \
+    static inline std::shared_ptr<DataType> type_constructor() { return Constructor(); } \
+  };
```

Comment:
I thought about that but that would lead to a cyclic include between `type_traits.h` and `type.h` thus I made those new traits.